File: src/cli/tools/design.go
// src/cli/tools/design.go
package tools

import (
	"fmt"
	"io"
	"os"
	"path/filepath"

	"github.com/lucas/oraculo/src/db"
	"github.com/lucas/oraculo/src/output"
	"github.com/spf13/cobra"
)

// newDesignCmd returns the "design" parent command with all subcommands.
func newDesignCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "design",
		Short: "Manage design artifacts",
	}
	cmd.AddCommand(
		newDesignSaveCmd(),
		newDesignGetCmd(),
	)
	return cmd
}

func newDesignSaveCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "save <story-name>",
		Short: "Save design markdown for a story (reads stdin)",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			name := args[0]
			w := cmd.OutOrStdout()

			database, epicName, epicID, err := resolveEpic(cmd)
			if err != nil {
				return err
			}

			// Verify the story exists.
			store := db.NewStoryStore(database)
			if _, err := store.GetByName(epicID, name); err != nil {
				output.WriteError(w, err)
				return err
			}

			content, err := io.ReadAll(cmd.InOrStdin())
			if err != nil {
				output.WriteError(w, fmt.Errorf("read stdin: %w", err))
				return err
			}

			dir := filepath.Join(".oraculo", "epics", epicName, "stories", name)
			if err := os.MkdirAll(dir, 0o755); err != nil {
				output.WriteError(w, fmt.Errorf("create directory: %w", err))
				return err
			}

			path := filepath.Join(dir, "design.md")
			if err := os.WriteFile(path, content, 0o644); err != nil {
				output.WriteError(w, fmt.Errorf("write file: %w", err))
				return err
			}

			return output.WriteJSON(w, map[string]any{
				"name":  name,
				"epic":  epicName,
				"path":  path,
				"saved": true,
			})
		},
	}
	cmd.Flags().String("epic", "", "Parent epic name (required)")
	_ = cmd.MarkFlagRequired("epic")
	return cmd
}

func newDesignGetCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "get <story-name>",
		Short: "Get design markdown for a story",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			name := args[0]
			w := cmd.OutOrStdout()

			_, epicName, _, err := resolveEpic(cmd)
			if err != nil {
				return err
			}

			path := filepath.Join(".oraculo", "epics", epicName, "stories", name, "design.md")
			content, err := os.ReadFile(path)
			if err != nil {
				wrappedErr := fmt.Errorf("story %q design: %w", name, err)
				output.WriteError(w, wrappedErr)
				return wrappedErr
			}

			_, err = w.Write(content)
			return err
		},
	}
	cmd.Flags().String("epic", "", "Parent epic name (required)")
	_ = cmd.MarkFlagRequired("epic")
	return cmd
}

File: apps/orchestrator/src/cli/tools/design.ts
import { Command } from "commander";
import { join } from "node:path";
import { mkdirSync } from "node:fs";
import { EpicStore } from "../../db/stores/epic-store.ts";
import { StoryStore } from "../../db/stores/story-store.ts";
import { writeJSON, writeError } from "../../utils/output.ts";
import { dbFromCmd } from "../context.ts";

export function designCommands(): Command {
  const cmd = new Command("design").description("Manage design artifacts");

  cmd.addCommand(designSaveCmd());
  cmd.addCommand(designGetCmd());

  return cmd;
}

function designSaveCmd(): Command {
  return new Command("save")
    .description("Save design markdown for a story (reads stdin)")
    .argument("<story-name>", "Story name")
    .requiredOption("--epic <name>", "Parent epic name")
    .action(async (name: string, opts: any, cmd: Command) => {
      try {
        const db = dbFromCmd(cmd);
        const epicStore = new EpicStore(db);
        const epic = epicStore.getByName(opts.epic);

        // Verify story exists
        const storyStore = new StoryStore(db);
        storyStore.getByName(epic.id, name);

        const content = await Bun.stdin.text();
        const dir = join(
          ".oraculo",
          "epics",
          opts.epic,
          "stories",
          name,
        );
        mkdirSync(dir, { recursive: true });
        const path = join(dir, "design.md");
        await Bun.write(path, content);

        writeJSON({ name, epic: opts.epic, path, saved: true });
      } catch (err) {
        writeError(err);
      }
    });
}

function designGetCmd(): Command {
  return new Command("get")
    .description("Get design markdown for a story")
    .argument("<story-name>", "Story name")
    .requiredOption("--epic <name>", "Parent epic name")
    .action(async (name: string, opts: any, cmd: Command) => {
      try {
        const db = dbFromCmd(cmd);
        const epicStore = new EpicStore(db);
        epicStore.getByName(opts.epic); // validate epic exists

        const path = join(
          ".oraculo",
          "epics",
          opts.epic,
          "stories",
          name,
          "design.md",
        );
        const file = Bun.file(path);
        if (!(await file.exists())) {
          writeError(
            new Error(`story "${name}" design: file not found`),
          );
          return;
        }
        process.stdout.write(await file.text());
      } catch (err) {
        writeError(err);
      }
    });
}
